File: python/src/wata/digital_goods/stars.py
"""Telegram Stars (токен терминала Stars): `/api/stars/...`.

Количество звёзд должно быть в диапазоне 50–50000. Заказы дороже порога
модерации попадают в статус `Review` и **не выполняются**, пока их явно не
подтвердят через :meth:`confirm_order` — иначе интегратор будет ждать выдачи,
которой не будет.
"""

from __future__ import annotations

from typing import Any

from .._http import AsyncTransport, SyncTransport
from ..types import DgObject, StarsOrderStatus, clean
from ._common import PREFIX, require_async, require_sync

_PRODUCT = "stars"
MIN_COUNT = 50
MAX_COUNT = 50000


def _validate_count(count: int) -> None:
    if not (MIN_COUNT <= count <= MAX_COUNT):
        raise ValueError(
            f"Количество звёзд должно быть в диапазоне {MIN_COUNT}..{MAX_COUNT}, получено {count}"
        )


def _create_body(
    *,
    username: str,
    count: int,
    amount: float,
    description: str,
    order_id: str,
    extra: dict[str, Any],
) -> dict[str, Any]:
    """Тело `POST /stars`. Имена полей — как в API, camelCase."""

    _validate_count(count)
    return clean(
        {
            "username": username,
            "count": count,
            "amount": amount,
            "description": description,
            "orderId": order_id,
            **extra,
        }
    )


class StarsAPI:
    """Синхронный доступ к Telegram Stars."""

    def __init__(self, transport: SyncTransport | None) -> None:
        self._transport = transport

    def price(self, *, username: str, **extra: Any) -> DgObject:
        """`GET /stars/price` — стоимость звезды и минимальная сумма заказа.

        Ответ содержит ``starPrice`` и ``minPrice``.
        """

        transport = require_sync(self._transport, _PRODUCT)
        params = clean({"username": username, **extra})
        response = transport.request("GET", f"{PREFIX}/stars/price", params=params)
        return DgObject.from_api(response.json())

    def create(
        self,
        *,
        username: str,
        count: int,
        amount: float,
        description: str,
        order_id: str,
        **extra: Any,
    ) -> DgObject:
        """`POST /stars` — создать заказ на покупку звёзд.

        :param username: получатель в Telegram
        :param count: количество звёзд, 50–50000
        :param amount: сумма заказа
        :param description: описание заказа
        :param order_id: ваш идентификатор заказа

        Заказ дороже порога модерации возвращается в статусе ``Review`` и
        требует явного :meth:`confirm_order` либо :meth:`reject_order`.
        В ответе приходят ``price``, ``commission`` и ``paymentLink``.
        """

        transport = require_sync(self._transport, _PRODUCT)
        body = _create_body(
            username=username,
            count=count,
            amount=amount,
            description=description,
            order_id=order_id,
            extra=extra,
        )
        response = transport.request("POST", f"{PREFIX}/stars", json_body=body, mutating=True)
        return DgObject.from_api(response.json(), status_enum=StarsOrderStatus)

    def get_order(self, order_id: str) -> DgObject:
        """`GET /stars/order/{id}`."""

        transport = require_sync(self._transport, _PRODUCT)
        response = transport.request("GET", f"{PREFIX}/stars/order/{order_id}")
        return DgObject.from_api(response.json(), status_enum=StarsOrderStatus)

    def confirm_order(self, order_id: str) -> DgObject:
        """`POST /stars/order/{id}/confirm` — подтвердить заказ в статусе `Review`."""

        transport = require_sync(self._transport, _PRODUCT)
        response = transport.request(
            "POST", f"{PREFIX}/stars/order/{order_id}/confirm", mutating=True
        )
        return DgObject.from_api(response.json(), status_enum=StarsOrderStatus)

    def reject_order(self, order_id: str) -> DgObject:
        """`POST /stars/order/{id}/reject` — отклонить заказ в статусе `Review`."""

        transport = require_sync(self._transport, _PRODUCT)
        response = transport.request(
            "POST", f"{PREFIX}/stars/order/{order_id}/reject", mutating=True
        )
        return DgObject.from_api(response.json(), status_enum=StarsOrderStatus)


class AsyncStarsAPI:
    """Асинхронный доступ к Telegram Stars — та же логика, что и :class:`StarsAPI`."""

    def __init__(self, transport: AsyncTransport | None) -> None:
        self._transport = transport

    async def price(self, *, username: str, **extra: Any) -> DgObject:
        """`GET /stars/price` — стоимость звезды и минимальная сумма заказа."""

        transport = require_async(self._transport, _PRODUCT)
        params = clean({"username": username, **extra})
        response = await transport.request("GET", f"{PREFIX}/stars/price", params=params)
        return DgObject.from_api(response.json())

    async def create(
        self,
        *,
        username: str,
        count: int,
        amount: float,
        description: str,
        order_id: str,
        **extra: Any,
    ) -> DgObject:
        """`POST /stars` — создать заказ на покупку звёзд."""

        transport = require_async(self._transport, _PRODUCT)
        body = _create_body(
            username=username,
            count=count,
            amount=amount,
            description=description,
            order_id=order_id,
            extra=extra,
        )
        response = await transport.request(
            "POST", f"{PREFIX}/stars", json_body=body, mutating=True
        )
        return DgObject.from_api(response.json(), status_enum=StarsOrderStatus)

    async def get_order(self, order_id: str) -> DgObject:
        """`GET /stars/order/{id}`."""

        transport = require_async(self._transport, _PRODUCT)
        response = await transport.request("GET", f"{PREFIX}/stars/order/{order_id}")
        return DgObject.from_api(response.json(), status_enum=StarsOrderStatus)

    async def confirm_order(self, order_id: str) -> DgObject:
        """`POST /stars/order/{id}/confirm` — подтвердить заказ в статусе `Review`."""

        transport = require_async(self._transport, _PRODUCT)
        response = await transport.request(
            "POST", f"{PREFIX}/stars/order/{order_id}/confirm", mutating=True
        )
        return DgObject.from_api(response.json(), status_enum=StarsOrderStatus)

    async def reject_order(self, order_id: str) -> DgObject:
        """`POST /stars/order/{id}/reject` — отклонить заказ в статусе `Review`."""

        transport = require_async(self._transport, _PRODUCT)
        response = await transport.request(
            "POST", f"{PREFIX}/stars/order/{order_id}/reject", mutating=True
        )
        return DgObject.from_api(response.json(), status_enum=StarsOrderStatus)

File: node/src/digital-goods/stars.ts
import { WataConfigError } from '../errors.js';
import { Transport, mapQuery } from '../http.js';
import type { StarsOrderStatus } from '../types.js';

const MIN_COUNT = 50;
const MAX_COUNT = 50000;

export interface StarsPriceQuery {
  /** Telegram username получателя. */
  username: string;
}

export interface StarsPrice {
  starPrice: number;
  minPrice: number;
  raw: unknown;
}

export interface CreateStarsOrderRequest {
  username: string;
  /** Количество звёзд, от 50 до 50000. */
  count: number;
  amount: number;
  description: string;
  orderId: string;
}

export interface StarsOrderCreated {
  orderId: string;
  username: string;
  count: number;
  amount: number;
  price: number;
  commission: number;
  description: string;
  paymentLink: string;
  raw: unknown;
}

export interface StarsOrderStatusResult {
  /**
   * Статус заказа. Заказы дороже порога (порядка 2000 ₽) попадают в
   * статус `Review` и НЕ будут выполнены сами по себе — их нужно явно
   * подтвердить через `confirm()`, иначе покупатель никогда не
   * получит звёзды.
   */
  status: StarsOrderStatus;
  username: string;
  count: number;
  amount: number;
  description?: string;
  creationTime?: string;
  raw: unknown;
}

function assertValidCount(count: number): void {
  if (!Number.isFinite(count) || count < MIN_COUNT || count > MAX_COUNT) {
    throw new WataConfigError(`Количество звёзд должно быть от ${MIN_COUNT} до ${MAX_COUNT}, получено: ${count}.`);
  }
}

/** Продукт "Telegram Stars" (собственный терминал, собственный токен, отдельный от Steam). */
export class StarsApi {
  constructor(private readonly transport: Transport) {}

  /** Стоимость звезды и минимально допустимая сумма для указанного получателя. */
  price(query: StarsPriceQuery): Promise<StarsPrice> {
    return this.transport.requestWithRaw<StarsPrice>({ method: 'GET', path: '/stars/price', query: mapQuery(query, {}) });
  }

  /**
   * Создать заказ. Не повторяется автоматически при сбое. Заказы
   * дороже порога попадут в статус `Review` — см. `confirm()`/`reject()`.
   */
  create(request: CreateStarsOrderRequest): Promise<StarsOrderCreated> {
    assertValidCount(request.count);
    return this.transport.requestWithRaw<StarsOrderCreated>({
      method: 'POST',
      path: '/stars',
      body: request,
      idempotent: false,
    });
  }

  /** Статус заказа. */
  order(id: string): Promise<StarsOrderStatusResult> {
    return this.transport.requestWithRaw<StarsOrderStatusResult>({
      method: 'GET',
      path: `/stars/order/${encodeURIComponent(id)}`,
    });
  }

  /** Подтвердить заказ, находящийся в статусе `Review` (переводит его в `Paid`), чтобы он был выполнен. */
  confirm(id: string): Promise<StarsOrderStatusResult> {
    return this.transport.requestWithRaw<StarsOrderStatusResult>({
      method: 'POST',
      path: `/stars/order/${encodeURIComponent(id)}/confirm`,
      idempotent: false,
    });
  }

  /** Отклонить заказ, находящийся в статусе `Review` (переводит его в `Refunded`). */
  reject(id: string): Promise<StarsOrderStatusResult> {
    return this.transport.requestWithRaw<StarsOrderStatusResult>({
      method: 'POST',
      path: `/stars/order/${encodeURIComponent(id)}/reject`,
      idempotent: false,
    });
  }
}
